### PR TITLE
Redesign sidebar for dashboard and search

### PR DESCRIPTION
**Pull request by kmerz**
## Description
Instead of a big panel on the right side, the search and dashboard information will no be stored
in collapsible sidebar.
For that the SearchPage and the DashboardPage will be displayed in a new AppRouterPage
without  margin or paddings or footer. That way we can use the entire space for the search
and the sidebar.
The page will be rendered in a css-grid-layout:
```
--------------------------------------------
|      |     |                            |
|  1   |  2  |              3             |
|      |     |                            |
--------------------------------------------
```
The sidebar will be rendered collapsed in collumn 1 and un-collapsed in 1 and 2.
The widget area will be rendered in 2 and 3.
That way we get a overlaping effect. This design was chosen, since re-rendering  the
widget area can take seconds with lot's of widgets on them.

## Motivation and Context
We want to give the widget area as much space as possible. So the user can arrange the 
widgets more freely. 

## Screenshots (if appropriate):
Graylog (13)
Graylog (12)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: graylog2-web-interface/src/components/common/ReactGridContainer.jsx [hunk 1]
```diff
@@ -140,10 +148,19 @@ class ReactGridContainer extends React.Component {
...
+    super(props);
+    this.state = {
+      layout: this.computeLayout(props.positions),
+    };
+  }
+
   componentWillReceiveProps(nextProps) {
-    if (!lodash.isEqual(nextProps.positions, this.props.positions)) {
+    const { positions } = this.props;
+    if (!lodash.isEqual(nextProps.positions, positions)) {
```
[the remark this answers is not in the document]
Comment (kmerz) [reply graylog2-web-interface/src/components/common/ReactGridContainer.jsx]: :+1:

File: graylog2-web-interface/src/routing/AppWithExtendedSearchBar.jsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+const CurrentUserStore = StoreProvider.getStore('CurrentUser');
+const SearchStore = StoreProvider.getStore('Search');
+const SavedSearchesStore = StoreProvider.getStore('SavedSearches');
+const StreamsStore = StoreProvider.getStore('Streams');
+const ConfigurationsStore = StoreProvider.getStore('Configurations');
+const SavedSearchesActions = ActionsProvider.getActions('SavedSearches');
+const ConfigurationActions = ActionsProvider.getActions('Configuration');
+
+const AppWithExtendedSearchBar = createReactClass({
```
[the remark this answers is not in the document]
Comment (kmerz) [reply graylog2-web-interface/src/routing/AppWithExtendedSearchBar.jsx]: Already done :+1:

File: graylog2-web-interface/src/routing/AppWithExtendedSearchBar.jsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  propTypes: {
+    children: PropTypes.element.isRequired,
+    params: PropTypes.shape({
+      streamId: PropTypes.string,
+    }).isRequired,
+  },
+
+  mixins: [
+    Reflux.connect(CurrentUserStore),
```
[the remark this answers is not in the document]
Comment (kmerz) [reply graylog2-web-interface/src/routing/AppWithExtendedSearchBar.jsx]: Not needed actually :+1:

File: graylog2-web-interface/src/views/components/WidgetGrid.jsx [hunk 4]
```diff
@@ -136,14 +137,18 @@ class WidgetGrid extends React.Component {
                           useDragHandle=".widget-drag-handle">
         {widgets}
       </ReactGridContainer>
-    ) : null;
+    ) : <span />;
```
[the remark this answers is not in the document]
Comment (kmerz) [reply graylog2-web-interface/src/views/components/WidgetGrid.jsx]: to fix a test: ```● <WidgetGrid /> › should render with minimal props

    Invariant Violation: React.cloneElement(...): The argument must be a React element, but you passed null.

      144 |           return (
      145 |             <div className="dashboard">
    > 146 |               {React.cloneElement(grid, { width: size.width })}
          |                      ^
      147 |               {staticWidgets}
      148 |             </div>
      149 |           );

      at invariant (node_modules/react/cjs/react.development.js:88:15)
      at Object.cloneElement (node_modules/react/cjs/react.development.js:840:51)
      at Object.cloneElementWithValidation [as cloneElement] (node_modules/react/cjs/react.development.js:1782:33)
      at cloneElement (src/views/components/WidgetGrid.jsx:146:22)
      at children (src/views/components/WidgetGrid.test.jsx:29:35)
      at renderWithHooks (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:5036:18)
      at mountIndeterminateComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7118:13)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7723:16)
      at performUnitOfWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11413:12)
      at workLoop (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11445:24)
      at renderRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11528:7)
      at performWorkOnRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12416:7)
      at performWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12328:7)
      at performSyncWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12302:3)
      at requestWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12171:5)
      at scheduleWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11986:5)
      at scheduleRootUpdate (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12585:3)
      at updateContainerAtExpirationTime (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12613:10)
      at updateContainer (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12624:10)
      at Object.create (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:13057:5)
      at Object.create (src/views/components/WidgetGrid.test.jsx:34:30)
```

[the remark this answers is not in the document]
Comment (kmerz) [reply graylog2-web-interface/src/views/components/WidgetGrid.jsx]: Well not quite true, to handle empty widgets. The tests only shows it. Can't call React.cloneElement with null.

File: graylog2-web-interface/src/views/components/sidebar/SideBar.css [hunk 5]
```diff
@@ -1,10 +1,74 @@
...
+    white-space: nowrap;
+}
+
+:local(.openFieldContent) {
+    white-space: nowrap;
+    padding: 20px;
+    position: absolute !important;
+    bottom: 20px;
+    height: calc(100vh - 314px);
+    max-height: calc(100vh - 314px);
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/components/sidebar/SideBar.css]: Is there no other way than doing pixel-based calculation?

I guess we can expect this to break/need adaption when we are changing anything that implicitly changes the size of the elements going into this calculation?

File: graylog2-web-interface/src/views/pages/ExtendedSearchPage.test.jsx [hunk 6]
```diff
@@ -60,8 +99,7 @@ describe('ExtendedSearchPage', () => {
     }));
     FieldTypesActions.all = mockAction(jest.fn());
     SearchMetadataActions.parseSearch = mockAction(jest.fn(() => mockPromise(SearchMetadata.empty())));
-    SearchMetadataStore.listen = jest.fn(() => jest.fn());
-    // $FlowFixMe: Somehow flow does not see the `listen` property.
+    SearchMetadataStore.listen = jest.fn(() => jest.fn()); // $FlowFixMe: Somehow flow does not see the `listen` property.
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/pages/ExtendedSearchPage.test.jsx]: I think a linebreak got lost.

**Pull request merged**